---
layout: archive
title: "GPLVM Presentation"
permalink: /gplvm.html
author_profile: true
---

<!DOCTYPE html>
<html lang="en"><head>
<script src="presentation_files/libs/clipboard/clipboard.min.js"></script>
<script src="presentation_files/libs/quarto-html/tabby.min.js"></script>
<script src="presentation_files/libs/quarto-html/popper.min.js"></script>
<script src="presentation_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="presentation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="presentation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.548">

  <meta name="author" content="Ruby Sedgwick">
  <title>Gaussian Process Latent Variable Models</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/theme/quarto.css">
  <link rel="stylesheet" href="styles.css">
  <link href="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="images/imperial_logo.jpg," data-background-opacity="0.8" data-background-position="right 25px top 25px" data-background-size="20%, 20%" class="quarto-title-block center">
  <h1 class="title">Gaussian Process Latent Variable Models</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Ruby Sedgwick 
</div>
<div class="quarto-title-author-email">
<a href="mailto:r.sedgwick19@imperial.ac.uk">r.sedgwick19@imperial.ac.uk</a>
</div>
        <p class="quarto-title-affiliation">
            Imperial College London
          </p>
    </div>
</div>

</section>
<section class="slide level2">

<!-- ## Motivation
\definecolor{causalcolour}{rgb}{0.00, 0 .0, 1.0}
\newcommand{\veryshortarrow}[1][3pt]{\mathrel{%
   \hbox{\rule[\dimexpr\fontdimen22\textfont2-.2pt\relax]{#1}{.4pt}}%
   \mkern-4mu\hbox{\usefont{U}{lasy}{m}{n}\symbol{41}}}}
\newcommand{\model}{\mathcal{M}}
\newcommand{\xcausey}{\model_{\textbf{X} \rightarrow Y}}
\newcommand{\ycausex}{\model_{X \leftarrow \textbf{Y}}}
\newcommand{\nomodelxcausey}{\textbf{\color{causalcolour}X} \rightarrow Y}
\newcommand{\nomodelycausex}{X\leftarrow\textbf{\color{causalcolour}Y}}
\newcommand{\setmarginals}{\mathcal{R}}
\newcommand{\setconditionals}{\mathcal{C}}
\newcommand{\bfx}{\mathbf{x}}
\newcommand{\bfy}{\mathbf{y}}
\newcommand{\suchthat}{\,\,|\,\,}
\newcommand{\data}{\mathcal{D}}
\newcommand{\calcd}[1][]{\mathrm{d}#1\xspace}
\usepackage{tikz}
\usetikzlibrary{arrows,automata}
\usetikzlibrary{bayesnet}


::: {.fragment}
::: {style="margin-top: 1.0em"}
- Data is often high dimensional
:::  -->
</section>
<section id="overview" class="slide level2">
<h2>Overview</h2>
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Motivation.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>Principal component analysis.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Gaussian process latent variable models.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Applications.</li>
</ul>
</div>
</section>
<section id="motivation" class="slide level2">
<h2>Motivation</h2>
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Data is often high dimensional.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>In theory, in high dimensions we need a lot of data to learn anything (curse of dimensionality).</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Often the data actually lies on a lower dimensional manifold.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>This means we can use useful representations from relatively small amounts of data.</li>
</ul>
</div>
</section>
<section id="motivation-1" class="slide level2">
<h2>Motivation</h2>
<div style="margin-top: 1.0em">
<p>We want to learn a mapping from low dimensional space to a high dimensional space<br>
<span class="math display">\[
h \in \mathbb{R}^Q \mapsto y \in \mathbb{R}^D
\]</span></p>
<p>where <span class="math inline">\(y\)</span> is observed, <span class="math inline">\(h\)</span> is an unobserved latent variable, and <span class="math inline">\(D &gt; Q\)</span>.</p>
</div>
<div class="fragment">
<div style="margin-top: 0.0em">
<div style="display: flex; align-items: center; justify-content: center;">
<p><!-- First image --> <img src="images/3d_s_curve_flat.png" alt="3D S-Curve" style="width: 45%; height: auto ; margin-left: 250px;"></p>
<!-- Arrow in the middle -->
<div style="font-size: 3em; margin: 0 10px 0 5px;">
→
</div>
<p><!-- Second image --> <img src="images/3d_s_curve.png" alt="3D S-Curve Flat" style="width: 60%; height: auto;"></p>
</div>
</div>
</div>
<!-- :::: {.columns .middle}
::: {.column width="35%"}
::: {style="text-align: center"}
::: {.fragment}
Data is often high dimensional  
<img src="images/3d_s_curve.png" alt="drawing" height="300"/>

:::
:::
::: -->
<!-- ::: {.column width="0%"}
<!-- Empty column for spacing -->
<!-- :::  -->
<!-- 
::: {.column width="65%"}
::: {style="text-align: center"}
::: {.fragment}
Often the data actually lies on a low dimensional manifold  
<img src="images/3d_s_curve_flat.png" alt="drawing" height="250"/>

:::
:::
:::
::::  -->
</section>
<section id="motivating-example" class="slide level2">
<h2>Motivating Example</h2>
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Simulated Dataset, with 12 spatial dimensions, of pipeline transporting a mix of oil, water and gas<sup>1</sup>.</li>
</ul>
</div>
<div style="font-size: 0.6em">
<table style="width:100%;">
<colgroup>
<col style="width: 5%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
<th>11</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>0.6022</td>
<td>0.4851</td>
<td>0.9271</td>
<td>0.4647</td>
<td>1.1818</td>
<td>0.3103</td>
<td>1.7287</td>
<td>0.0006</td>
<td>2.1438</td>
<td>-0.1692</td>
<td>0.0210</td>
<td>1.7062</td>
</tr>
<tr class="even">
<td>1</td>
<td>0.7532</td>
<td>0.3612</td>
<td>0.3635</td>
<td>1.0189</td>
<td>0.3685</td>
<td>0.5868</td>
<td>0.3342</td>
<td>1.0666</td>
<td>0.2891</td>
<td>0.6552</td>
<td>0.3743</td>
<td>1.0184</td>
</tr>
<tr class="odd">
<td>2</td>
<td>0.5023</td>
<td>0.2296</td>
<td>0.8095</td>
<td>0.3391</td>
<td>0.8645</td>
<td>0.4546</td>
<td>0.7044</td>
<td>0.4620</td>
<td>0.8431</td>
<td>0.5025</td>
<td>0.7605</td>
<td>0.3984</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="fragment">
<ul>
<li>The flow in the pipe takes one of the three flow configurations: horizontally stratified, nested annular or homogeneous mixture flow.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>However, the data is inherently two dimensional as the flow is determined by the fractions of water and oil (fraction of air is then determined as it must sum to one).</li>
</ul>
</div>
<!-- ::: {.fragment}
::: {style="margin-top: 1.0em"}
- Single cell gene expression dataset[^1] 
:::
:::

[^1]:Guo, Guoji, et al. "Resolution of cell fate decisions revealed by single-cell gene expression analysis from zygote to blastocyst." Developmental cell 18.4 (2010): 675-685. available at: <https://github.com/sods/ods>

::: {.fragment}
- 48 genes measured at 437 time points
:::
::: {.fragment}
  -  However, we know there are 10 stages of development 
:::
::: {.fragment}
How can we interpret this data? 
::: -->
<aside><ol class="aside-footnotes"><li id="fn1"><p>Bishop, Christopher M., and Gwilym D. James. ‘Analysis of multiphase flows using dual-energy gamma densitometry and neural networks’. Nuclear Instruments and Methods in Physics Research Section A: Accelerators, Spectrometers, Detectors and Associated Equipment 327.2-3 (1993): 580-593.</p></li></ol></aside></section>
<section id="principal-component-analysis" class="slide level2">
<h2>Principal Component Analysis</h2>
<div class="fragment">
<!-- ::: {style="margin-top: -1.5em"} -->
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>Can we learn a low dimensional representation of this data?</p>
</div>
</div>
<div class="columns">
<div class="column" style="width:60%;">
<!-- ::: {style="margin-top: 1.0em"}  -->
<div class="fragment">
<ul>
<li>For a dataset <span class="math inline">\(Y = [\mathbf{y}_1, ..., \mathbf{y}_N]^T\)</span>, <span class="math inline">\(\mathbf{y}_n \in \mathbb{R}^{D}\)</span>,</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>We assume a linear mapping: <span class="math display">\[\mathbf{y}_n = \mathbf{W}\mathbf{h}_n,\]</span> where <span class="math inline">\(\mathbf{y}_n \in \mathbb{R}^D\)</span> is the observed data, <span class="math inline">\(\mathbf{h}_n \in \mathbb{R}^Q\)</span> is the corrosponding latent point and <span class="math inline">\(\mathbf{W} \in \mathbb{R}^{D \times Q}\)</span> is a linear mapping.</li>
</ul>
<!-- - $H=[\mathbf{h}_1, ..., \mathbf{h}_N]^T$,  $\mathbf{h}_n \in \mathbb{R}^{Q}$ are points in a low dimensional latent space.
- $Y = [\mathbf{y}_1, ..., \mathbf{y}_N]^T$, $\mathbf{y}_n \in \mathbb{R}^{P}$ are observations.
- $W \in \mathbb{R}^{D\times Q}$ is a linear mapping. -->
</div>
<div class="fragment">
<div style="margin-top: -1.0em">
<ul>
<li><span class="math inline">\(\mathbf{W}\)</span> can be calculated in closed form by finding the eigenvalues and vectors of the covariance matrix <span class="math inline">\(cov(\mathbf{Y},\mathbf{Y}).\)</span></li>
</ul>
<!-- - $H=[\mathbf{h}_1, ..., \mathbf{h}_N]^T$,  $\mathbf{h}_n \in \mathbb{R}^{Q}$ are points in a low dimensional latent space.
- $Y = [\mathbf{y}_1, ..., \mathbf{y}_N]^T$, $\mathbf{y}_n \in \mathbb{R}^{P}$ are observations.
- $W \in \mathbb{R}^{D\times Q}$ is a linear mapping. -->
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="fragment">
<div style="margin-top: 2.0em">
<!-- ::: {style="margin-top: 0.5em"}  -->
<figure style="text-align: center;">
<img src="images/PPCA_oil_flow.png" alt="drawing" width="500">
<figcaption>
PCA on the Oil Flow Dataset
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<div style="margin-top: -1.5em">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<div class="fragment">
<p>But real data is noisy…</p>
</div>
</div>
</div>
</section>
<section id="probabilistic-principal-component-analysis" class="slide level2">
<h2>Probabilistic Principal Component Analysis</h2>
<div class="fragment">
<ul>
<li>In Probabilistic PCA we assume the data in noisy:</li>
</ul>
</div>
<div class="fragment">
<div style="text-align: center">
<p><span class="math display">\[
\mathbf{y}_n = \mathbf{W}\mathbf{h}_n + \epsilon_n,
\]</span></p>
</div>
</div>
<div class="fragment">
<ul>
<li>where <span class="math inline">\(\epsilon_n\)</span> is a Gaussian noise term with variance <span class="math inline">\(\sigma^2\)</span>:</li>
</ul>
</div>
<div class="fragment">
<div style="text-align: center">
<p><span class="math display">\[
\epsilon_n \sim \mathcal{N}(\epsilon_n| \mathbf{0}, \sigma^2\mathbf{I}).
\]</span></p>
</div>
</div>
<div class="fragment">
<ul>
<li>This gives us the likelihood:</li>
</ul>
</div>
<div class="fragment">
<div style="text-align: center">
<p><span class="math display">\[
p(\mathbf{y}_n |\mathbf{h}_n, \mathbf{W}, \sigma^2) = \mathcal{N}(\mathbf{y}_n | \mathbf{W}\mathbf{h}_n, \sigma^2\mathbf{I}).
\]</span></p>
</div>
</div>
</section>
<section id="probabilistic-principal-component-analysis-1" class="slide level2">
<h2>Probabilistic Principal Component Analysis</h2>
<div class="fragment">
<ul>
<li><p>Normally in PPCA, the latents are marginalised out by applying the prior <span class="math inline">\(p(\mathbf{h}_n) = \mathcal{N}(\mathbf{0}, \mathbf{I})\)</span>, giving a marginal likelihood of:</p>
<p><span class="math display">\[\begin{align*}
  p(\mathbf{y}_n | \mathbf{h}_n, \mathbf{W}, \sigma^2) &amp; = \int p(\mathbf{y}_n |\mathbf{h}_n, \mathbf{W}, \sigma^2)p(\mathbf{h}_n) d\mathbf{h}_n  \\
  &amp; = \mathcal{N}(\mathbf{y}_n | \mathbf{0}, \mathbf{W}\mathbf{W}^T + \sigma^2\mathbf{I}).
  \end{align*}\]</span></p></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>However, we can instead marginalise out <span class="math inline">\(\mathbf{W}\)</span> using the prior:</li>
</ul>
<p><span class="math display">\[p(\mathbf{W}) = \prod_{i=1}^D\mathcal{N}(\mathbf{w_i}|\mathbf{0}, \mathbf{I}).\]</span></p>
</div>
</section>
<section id="probabilistic-principal-component-analysis-2" class="slide level2">
<h2>Probabilistic Principal Component Analysis</h2>
<div class="fragment">
<ul>
<li>For a set of data points <span class="math inline">\(Y = [\mathbf{y}_1, ..., \mathbf{y}_N]^T\)</span> and corropsonding latent variables <span class="math inline">\(H=[\mathbf{h}_1, ..., \mathbf{h}_N]^T\)</span>, using this prior allows the marginal likelihood to take the form: <span class="math display">\[p(Y|\mathbf{H}, \sigma^2) = \prod_{i=1}^D p(\mathbf{y}_{:,d}|\mathbf{H}, \sigma^2),\]</span> where: <span class="math display">\[p(\mathbf{y}_{:,d} |\mathbf{h}_{:,d}, \sigma^2) = \mathcal{N}(\mathbf{y}_{:,d} | \mathbf{0}, \mathbf{H}\mathbf{H}^T + \sigma^2\mathbf{I}).\]</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>We can then get <span class="math inline">\(\mathbf{W}\)</span> by optimising this marginal likelihood.</li>
</ul>
</div>
</section>
<section id="probabilistic-principal-component-analysis-3" class="slide level2">
<h2>Probabilistic Principal Component Analysis</h2>
<div class="fragment">
<figure style="text-align: center;">
<img src="images/PPCA_oil_flow.png" alt="drawing" width="400">
<figcaption>
PPCA on the Oil Flow Dataset
</figcaption>
</figure>
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>Probabilistic Principal Component Analysis gives us a linear mapping <span class="math inline">\(\mathbf{h} \mapsto \mathbf{y}\)</span>.</p>
</div>
</div>
<div class="fragment">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>Is there a way we can learn more complex mappings?</p>
</div>
</div>
</section>
<section id="gaussian-process-latent-variable-model-2" class="slide level2">
<h2>Gaussian Process Latent Variable Model <sup>1</sup></h2>
<div class="fragment">
<div class="columns">
<div class="column" style="width:55%;">
<div style="margin-top: 1.0em">
<ul>
<li>We can use a Gaussian process model but this time the inputs <span class="math inline">\(\mathbf{H}=[\mathbf{h}_1, ..., \mathbf{h}_N]^T\)</span> are <strong>not observed</strong>.</li>
</ul>
</div>
<ul>
<li>The outputs <span class="math inline">\(\mathbf{Y} = [\mathbf{y}_1, ..., \mathbf{y}_N]^T\)</span> are observed.</li>
</ul>
</div><div class="column" style="width:45%;">
<p><img src="images/gaussian_process.png" alt="drawing" height="250"></p>
</div>
</div>
</div>
<div class="fragment">
<div style="margin-top: -1.5em">
<ul>
<li>This gives us the generative model:</li>
</ul>
<p><span class="math display">\[
\mathbf{y} = f(\mathbf{h}) + \mathbf{\epsilon} ~~~~ f \sim \mathcal{GP}(\mathbf{0}, k(\mathbf{h}, \mathbf{h})) ~~~~\mathbf{\epsilon} \sim \mathcal{N}(0, \sigma_n^2).
\]</span></p>
</div>
</div>
<div class="fragment">
<div style="margin-top: -1.0em">
<ul>
<li><span class="math inline">\(\mathbf{h}_n \in \mathbb{R}^{Q}\)</span> and <span class="math inline">\(\mathbf{y}_n \in \mathbb{R}^{D}\)</span>, <span class="math inline">\(D&gt;Q\)</span> and <span class="math inline">\(\sigma_n^2\)</span> is the noise variance. Kernel <span class="math inline">\(k(\mathbf{h}, \mathbf{h})\)</span> is parameterised by hyperparameters <span class="math inline">\(\mathbf{\theta}\)</span>.</li>
</ul>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn2"><p>Lawrence, Neil. “Probabilistic non-linear principal component analysis with Gaussian process latent variable models.” Journal of machine learning research 6.11 (2005).</p></li></ol></aside></section>
<section id="gaussian-process-latent-variable-model" class="slide level2">
<h2>Gaussian Process Latent Variable Model</h2>
<div class="fragment">
<!-- ::: {style="margin-top: 1.0em"} -->
<ul>
<li>We find the optimal latent variables <span class="math inline">\(\mathbf{H}\)</span>, kernel hyperparameters <span class="math inline">\(\mathbf{\theta}\)</span> and noise variance <span class="math inline">\(\sigma_n^2\)</span> by optimising the log marginal likelihood.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>To do this we take the likelihood:</li>
</ul>
<!-- ::: {style="margin-top: -1.5em"}  -->
<p><span class="math display">\[
p(\mathbf{Y}|\mathbf{F}, \mathbf{H}, \mathbf{\theta}, \sigma_n^2),
\]</span></p>
</div>
<div class="fragment">
<ul>
<li>and marginalize out the function values <span class="math inline">\(\mathbf{F}=f(\mathbf{H})\)</span>:</li>
</ul>
<!-- ::: {style="margin-top: -1.5em"}  -->
<p><span class="math display">\[\begin{align*}
     p(\mathbf{Y}|\mathbf{H}, \mathbf{\theta}, \sigma_n^2) &amp;= \int p(\mathbf{Y}|\mathbf{F}, \mathbf{H}, \mathbf{\theta}, \sigma_n^2) p(\mathbf{F}|\mathbf{H},\mathbf{\theta}) dF \\
    &amp; = \mathcal{N}(0, K(\mathbf{H}, \mathbf{H}) + \sigma_n^2 \mathbf{I}).
    \end{align*}
\]</span></p>
</div>
</section>
<section id="gaussian-process-latent-variable-model-1" class="slide level2">
<h2>Gaussian Process Latent Variable Model</h2>
<div class="fragment">
<!-- ::: {style="margin-top: -1.5em"}  -->
<ul>
<li>This gives the log marginal likelihood: <span class="math display">\[
\begin{align*}
  \log p(\mathbf{Y}|\mathbf{H}, \mathbf{\theta}, \sigma_n^2) =  &amp; - \underbrace{\frac{1}{2} \text{Tr}(K_H+\sigma_n^2\mathbf{I})^{-1}\mathbf{Y}\mathbf{Y}^T}_{\text{ data fit}}  - \underbrace{\frac{D}{2} \log|K_H+\sigma_n^2\mathbf{I}| }_{\text{complexity penalty}} \\ &amp; - \underbrace{\frac{D N}{2}\log 2 \pi}_{\text{norm. constant}}.
  \end{align*}
\]</span> where <span class="math inline">\(K_H=K(\mathbf{H},\mathbf{H})\)</span>.</li>
</ul>
</div>
</section>
<section id="gaussian-process-latent-variable-model-3" class="slide level2">
<h2>Gaussian Process Latent Variable Model</h2>
<div class="fragment">
<ul>
<li>Our GPLVM marginal likelihood is:</li>
</ul>
<div style="margin-top: -1em">
<p><span class="math display">\[\begin{align*}
p(Y|H, \mathbf{\theta}, \sigma_n^2) = \mathcal{N}(0, K(H, H) + \sigma_n^2 \mathbf{I}).
\end{align*}
\]</span></p>
</div>
</div>
<div class="fragment">
<ul>
<li>What does this look a lot like?</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Our PPCA marginal likelihood is:</li>
</ul>
<div style="margin-top: -1.5em">
<p><span class="math display">\[p(Y|X, \sigma^2) = \prod_{i=1}^D p(\mathbf{y}_{:,d}|X, \sigma^2),\]</span></p>
</div>
<div style="margin-top: -1.5em">
<p>where: <span class="math display">\[p(\mathbf{y}_n |\mathbf{h}_n, \sigma^2) = \mathcal{N}(\mathbf{y}_n | \mathbf{0}, \mathbf{H}\mathbf{H}^T + \sigma^2\mathbf{I}).\]</span></p>
</div>
</div>
<div class="fragment">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>PPCA is equivalent to GPLVM with a linear kernel!</p>
</div>
</div>
</section>
<section id="ppca-vs-gplvm" class="slide level2">
<h2>PPCA vs GPLVM</h2>
<div class="columns">
<div class="column" style="width:45%;">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>PPCA</p>
</div>
<div style="margin: -1em; padding: -1em;">
<div style="text-align: center; margin: -1em;">
<p><img src="images/PPCA_oil_flow.png" alt="drawing" width="450"></p>
</div>
</div>
<div style="margin: 0; padding: 0;">
<div style="text-align: center; margin: 0;">
<p>Linear mapping</p>
</div>
</div>
<div style="font-size: 0.8em">
<p><span class="math display">\[p(\mathbf{y}_n |\mathbf{h}_n, \sigma^2) = \mathcal{N}(\mathbf{y}_n | \mathbf{0}, \mathbf{H}\mathbf{H}^T + \sigma^2\mathbf{I}).\]</span></p>
</div>
</div><div class="column" style="width:10%;">
<!-- empty column to create gap -->
</div><div class="column" style="width:45%;">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>GPLVM</p>
</div>
<div style="margin: -1em; padding: -1em;">
<div style="text-align: center; margin: -1em;">
<p><img src="images/GPLVM_oil_flow.png" alt="drawing" width="450"></p>
</div>
</div>
<div style="margin: 0; padding: 0;">
<div style="text-align: center; margin: 0;">
<p>Non-linear mapping</p>
</div>
</div>
<div style="font-size: 0.8em">
<p><span class="math display">\[\begin{align*}
p(Y|H, \mathbf{\theta}, \sigma_n^2) = \mathcal{N}(0, K(\mathbf{H}, \mathbf{H}) + \sigma_n^2 \mathbf{I}).
\end{align*}
\]</span></p>
</div>
</div>
</div>
<!-- ::: {.fragment}
::: {style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em"}
PPCA is equivalent to GPLVM with a linear kernel! 
:::
::: -->
</section>
<section id="limitations-of-gplvm" class="slide level2">
<h2>Limitations of GPLVM</h2>
<div class="columns">
<div class="column" style="width:45%;">
<div class="fragment">
<div style="text-align: center; margin-top:3.0em">
<ul>
<li>There is no good way to choose the dimensions of the latent space <span class="math inline">\(Q\)</span>.</li>
</ul>
</div>
</div>
<div class="fragment">
<!-- ::: {style="text-align: center; margin-top:1.0em"} -->
<ul>
<li>Latent variables <span class="math inline">\(H\)</span> are deterministic, which can lead to overfitting.</li>
</ul>
</div>
<div class="fragment">
<!-- ::: {style="text-align: center; margin-top:1.0em"} -->
<ul>
<li>There is a version of the GPLVM called the Bayesian GPLVM that addresses these issues.</li>
</ul>
</div>
</div><div class="column" style="width:10%;">
<!-- empty column to create gap -->
</div><div class="column" style="width:45%;">
<div style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em">
<p>GPLVM</p>
</div>
<div style="margin: -1em; padding: -1em;">
<div style="text-align: center; margin: -1em;">
<p><img src="images/GPLVM_oil_flow.png" alt="drawing" width="450"></p>
</div>
</div>
<div style="margin: 0; padding: 0;">
<div style="text-align: center; margin: 0;">
<p>Non-linear mapping</p>
</div>
</div>
<div style="font-size: 0.8em">
<p><span class="math display">\[\begin{align*}
p(Y|H, \mathbf{\theta}, \sigma_n^2) = \mathcal{N}(0, K(\mathbf{H}, \mathbf{H}) + \sigma_n^2 \mathbf{I}).
\end{align*}\]</span></p>
</div>
</div>
</div>
<!-- 
## Bayesian Gaussian Process Latent Variable Model [^3] 


:::  {.fragment}
:::: {.columns}
::: {.column width="60%"}
- What if we put a prior on $H$? 
    $$
    \mathbf{h}_n \sim \mathcal{N}(\mathbf{\mu}_{h_n}, \Sigma_{h_n})
    $$
:::

::: {.column width="40%"}
<img src="images/standard_normal.png" alt="drawing" width="300"/> 

:::
::::
:::

::: {.fragment} 
::: {style="margin-top: -1.0em"} 
- The marginal likelihood then becomes:
$$
p(Y|H, \mathbf{\theta}, \sigma_n) = \int p(Y|F, H, \mathbf{\theta}, \sigma_n) p(F|H, \mathbf{\theta}) p(H) dF dH.
$$
:::
:::
[^3]:Titsias, Michalis, and Neil D. Lawrence. "Bayesian Gaussian process latent variable model." Proceedings of the thirteenth international conference on artificial intelligence and statistics. JMLR Workshop and Conference Proceedings, 2010.


::: {.fragment}
::: {style="margin-top: -1.0em"} 
- However this is intractable! $H$ appears non-linearly in inverse of $K(H,H) + \sigma^2\mathbf{I}$.
:::
:::

::: {.fragment}
- To solve this problem we use a method called variational inference.
:::   -->
<!-- 
::: {.fragment}
- However the marginal like is now intractable as $H$ appears non-linearly in inverse of $K(H,H) + \sigma^2\mathbf{I}$.
:::

::: {.fragment}
- To solve this problem we are going to use a technique called variational inference.
::: -->
<!-- ## Variational Inference


:::: {.columns}
::: {.column width="60%"}
:::{style="margin-top:1.0em}
::: {.fragment}
- Variational inference is a useful technique for approximating a distribution that can't be calculated directly.
:::
:::
::: {.fragment}
- We define an approximate distribution $q(x)$ of a form that is tractable.
:::

:::
::: {.column width="40%"}
<img src="images/vi.png" alt="drawing" width="400"/> 

:::
::::

::: {.fragment}
::: {style="margin-top:-1.0em}
- We then minimise the distance between this approximate distribution $q(x)$ and the true distribution $p(x)$.
:::
:::
::: {.fragment}
- So to learn the true distribution we minimise the Kullback-Leibler divergence which is a measure of distance between two distributions:
$$KL(q(x)|| p(x)) = \int q(x) \log \frac{q(x)}{p(x)} dx.$$
::: -->
<!-- 
::: {.fragment}
:::: {.columns}
::: {.column width="60%"}

:::
::: {.column width="40%"}
<img src="images/vi_1.png" alt="drawing" width="300"/> 

:::
::::
:::


::: {.fragment}
:::: {.columns}
::: {.column width="60%"}

- We then minimise the distance between this approximate distribution $q(x)$ and the true distribution $p(x)$ using the 
the Kullback-Leibler divergence, which is a measure of distance between two distributions

- So to learn the true distribution we minimise the KL-divergence
$$KL(q(x)|| p(x)) = \int q(x) \log \frac{q(x)}{p(x)} dx $$
:::
::: {.column width="40%"}

<img src="images/vi.png" alt="drawing" width="300"/> 

:::
::::
::: -->
<!-- ::: {.fragment}
- We define an approximate distribution $q(x)$ of a form that is tractable 
::: -->
<!-- ::: {.fragment}
- We then minimise the distance between this approximate distribution $q(x)$ and the true distribution $p(x)$
:::

::: {.fragment}
- We use the Kullback-Leibler divergence, which is a measure of distance between two distributions
:::

::: {.fragment}
- So to learn the true distribution we minimise the KL-divergence
$$KL(q(x)|| p(x)) = \int q(x) \log \frac{q(x)}{p(x)} dx $$
:::  -->
<!-- 
## Bayesian GPLVM Variational Inference

::: {.fragment}
- Our marginal likelihood is intractable:
$$p(Y|H, \mathbf{\theta}, \sigma_n) = \int p(Y|F, H, \mathbf{\theta}, \sigma_n) p(F|H, \mathbf{\theta}) p(H) dF dH.$$

:::
::: {.fragment}
- To address this we will use a variational distribution over the latent variables: 
$$q(H) = \prod^N_{n=1} \mathcal{N}(\mathbf{h_n}| M_n, S_n), $$
where $M_n$ and $S_n$ are the mean a variance of the variational distribution $q(\mathbf{h}_n)$.
:::

## Bayesian GPLVM Variational Inference

::: {.fragment}
- A lower bound can then be expressed via Jensen's inequality ($\mathbb{E}[g(x)] \geq g(\mathbb{E}[x])$) as follows:
$$\begin{align*}
    \log p(Y|H) &= \log \int p(Y|H) p(H) dH \\
     & = \log \int p(Y|H) p(H) \frac{q(H)}{q(H)} dH \\
     & \geq \int q(H) \log \frac{p(Y|H) p(H)}{q(H)} dH.
\end{align*}$$
:::
::: {.fragment}
::: {style="margin-top:-2.0em}
This can be rearranged to: 
$$
    \log p(Y|H) \geq \underbrace{\int q(H) \log p(Y|H) dH}_{\text{still intractable!}} - \underbrace{\int q(H) \log \frac{ q(H)}{p(H)} dH}_{\text{easy to calculate}}.
    \label{eq:BGPLVM_LB}
$$
:::
::: -->
<!-- ## Bayesian GPLVM VI 

::: {.fragment}
- we make $\int q(H) \log p(Y|H) dH$  by including inducing points (like in sparse Gaussian processes). After some manipulation we get the lower bound:
:::

$$\begin{multline*}
    \log p(Y|H) \geq  \int q(H) \Biggl[ \int q(U) \Biggl( \int p(F| U, H) \log p(Y|F) ~d(F) \\  + \log\frac{p(U}{q(U)} \Biggr) ~ d(U)    - \log \frac{ q(H)}{p(H)} \Biggr] dH.
\end{multline*}
$$

- So by maximising our lower bound, we can maximise our marginal likelihood  -->
<!-- ## Bayesian GPLVM VI 

- How do we make $\int q(H) \log p(Y|H) dH$ tractable?
- Add $N_U$ inducing variables $U = \in \mathbb{R}^{N_U \times D}$ evaluated at inducing point locations $Z = [\mathbf{z}_1, ..., \mathbf{z}_{N_U}]^T$ where $Z \in \mathbb{R}^{N_U \times Q}$
- This gives a joint Gaussian process prior between $F$ and $U$ as:
$$ p(F, U | H, Z) = p(F|U, H, Z) p(U |Z),
$$
- The joint distribution is then:
$$p(Y, F, U | H, Z) = p(Y| F)p(F|U, H, Z) p(U |Z). $$
- Conditioning on $U$ decouples $F$ from each other meaning: 
 $$p(Y|H) = \prod_{n=1}^N p(\mathbf{y_n}|H).$$ -->
<!-- ## Bayesian GPLVM VI 

Include illistration of inducing variables / points

- How do we make $\int q(H) \log p(Y|H) dH$ tractable?
- Add $N_U$ inducing variables $U = \in \mathbb{R}^{N_U \times D}$ evaluated at inducing point locations $Z = [\mathbf{z}_1, ..., \mathbf{z}_{N_U}]^T$ where $Z \in \mathbb{R}^{N_U \times Q}$
- This gives a joint Gaussian process prior between $F$ and $U$ as:
$$ p(F, U | H, Z) = p(F|U, H, Z) p(U |Z),$$
- The joint distribution is then:
$$ p(Y, F, U | H, Z) = p(Y| F)p(F|U, H, Z) p(U |Z). $$

## Bayesian GPLVM VI  -->
<!-- 
- Conditioning on $U$ decouples $F$ from each other meaning: 
 $$p(Y|H) = \prod_{n=1}^N p(\mathbf{y_n}|H).$$
- This means $p(Y)$ can be found by marginalising out $F$ and $U$ for any value of $Z$.

## Bayesian GPLVM VI  -->
<!-- 
- By introducing a second variational distribution:
 $$q(F, U) = p(F| U, H) q(U),$$
 we can get a tractable lower bound:
 $$
 \begin{align*}
    \log p(Y|H) &= \log  \int p(Y|F) p(F |U, H) p(U) dH \\
    &= \log \int p(Y|F) p(F |U, H) p(U) \frac{p(F |U, H) q(U)}{p(F |U, H) q(U)} dH \\
    &= \log \mathbb{E} \left[\frac{p(Y| F) p(F |U, H) p(U)}{p(F |U, H) q(U)}\right]_{p(F |U, H) q(U)} \\
    &\geq \int p(F |U, H) q(U) \log \left( \frac{p(Y|F) p(U)}{q(U)}  \right) ~ dF dU \\
    &= \int q(U) \left( \int p(F| U, H) \log p(Y|F) ~d(F) + \log\frac{p(U)}{q(U)} \right) ~ dU, 
  \end{align*}
  $$ -->
<!-- 
## Bayesian GPLVM VI 

- Putting this back into our lower bound we get: 
$$\begin{multline*}
    \log p(Y|H) \geq  \int q(H) \Biggl[ \int q(U) \Biggl( \int p(F| U, H) \log p(Y|F) ~d(F) \\  + \log\frac{p(U}{q(U)} \Biggr) ~ d(U)    - \log \frac{ q(H)}{p(H)} \Biggr] dH.
\end{multline*}
$$

- This is tractable! We therefore can maximise this lower bound instead of maximising the marginal likelihood.  -->
<!-- ## The Evolution of GPLVMs

::: {style="margin-top: 4em"} 
![](images/GPLVMs_of.png)
::: -->
<!-- ## BGPLVM Automatic Dimensionality reduction

:::  {.fragment}
- If we use an automatic relevance determining (ARD) kernel then the BGPLVM can automatically choose the dimensions. These kernels take the form:
$$K(H,H) = \prod_{d=1}^D k(\mathbf{h}_{:,d},\mathbf{h}_{:,d})$$
:::

- Plots of inverse lengthscales for GPLVM and BGPLVM with 5D for gene dataset.  -->
</section>
<section id="use-cases" class="slide level2">
<h2>Use Cases</h2>
<div class="fragment">
<div style="margin-top: 5.0em">
<ul>
<li>Learning pseudo-time in single cell data.</li>
</ul>
</div>
</div>
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Learning latent dynamics of motion capture data.</li>
</ul>
</div>
</div>
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Multi-output Gaussian processes.</li>
</ul>
</div>
</div>
</section>
<section id="psuedo-time-in-single-cell-data" class="slide level2">
<h2>Psuedo-time in Single Cell Data</h2>
<div class="columns">
<div class="column" style="width:45%;">
<div class="fragment">
<div style="margin-top: 1.0em">
<ul>
<li>Single cell gene expression dataset<sup>1</sup>.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>48 genes measured at 437 time points (i.e.&nbsp;<span class="math inline">\(Y\in\mathbb{R}^{48 \times 437}\)</span>).</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>However, we know there are 10 stages of development with a branching pattern.</li>
</ul>
</div>
</div><div class="column" style="width:10%;">
<!-- empty column to create gap -->
</div><div class="column" style="width:45%;">
<p>Cell Differentiation Hierarchy</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="presentation_files/figure-revealjs/unnamed-chunk-1-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<!-- ::: {style="margin-top: 1.0em"} -->
<div style="text-align: center; color: #107895; font-size: 1em; font-weight: 900em">
<p>Can we use a GPLVM to distinguish the different stages of cell developement?</p>
</div>
</div>
<aside><ol class="aside-footnotes"><li id="fn3"><p>Guo, Guoji, et al.&nbsp;“Resolution of cell fate decisions revealed by single-cell gene expression analysis from zygote to blastocyst.” Developmental cell 18.4 (2010): 675-685. available at: <a href="https://github.com/sods/ods" class="uri">https://github.com/sods/ods</a></p></li></ol></aside></section>
<section id="psuedo-time-in-cell-lines-5" class="slide level2">
<h2>Psuedo-time in Cell lines <sup>1</sup></h2>
<div class="fragment">
<ol type="1">
<li>Take observed data <span class="math inline">\(Y\in\mathbb{R}^{48 \times 437}\)</span>.</li>
</ol>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:40%;">
<!-- ::: {style="margin-top: -2.0em"} -->
<figure style="text-align: center;">
<p><img src="images/PCA_cell_data.png" alt="drawing" width="350"></p>
</figure>
<!-- ::: -->
</div><div class="column" style="width:60%;">
<ol start="2" type="1">
<li>Use the real time <span class="math inline">\(\mathbf{\tau} \in \mathbb{R}^{N}\)</span> to initialise the first dimension of the latent variable: <span class="math display">\[p(\mathbf{h}_{n,0}) =\mathbf{\tau}_{n,0}.\]</span></li>
</ol>
</div>
</div>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:60%;">
<ol start="3" type="1">
<li>Train the model by maximising the log marginal likelihood.</li>
</ol>
</div><div class="column" style="width:40%;">
<div style="margin-top: -2.0em">
<figure style="text-align: center;">
<p><img src="images/lml.png" alt="drawing" width="350"></p>
</figure>
</div>
</div>
</div>
</div>
<!-- 
::: {.fragment}
3. Train the model using the marginal likelihood

:::

:::

::: {.column width="10%"}
<!-- empty column to create gap -->
<!-- :::
::: {.column width="45%"}

<img src="images/BGPLVM_cell_data.png" alt="drawing" width="500"/> 

:::
:::: -->
<aside><ol class="aside-footnotes"><li id="fn4"><p>Ahmed, Sumon, Magnus Rattray, and Alexis Boukouvalas. “GrandPrix: scaling up the Bayesian GPLVM for single-cell data.” Bioinformatics 35.1 (2019): 47-54.</p></li></ol></aside></section>
<section id="psuedo-time-in-cell-lines" class="slide level2">
<h2>Psuedo-time in Cell lines</h2>
<div style="margin-top: 2.0em">
<figure style="text-align: center;">
<img src="images/GPLVM_cell_data.png" alt="drawing" height="450">
<figcaption>
GPLVM fitted to single-cell data.
</figcaption>
</figure>
</div>
<!-- [^5]:Ahmed, Sumon, Magnus Rattray, and Alexis Boukouvalas. "GrandPrix: scaling up the Bayesian GPLVM for single-cell data." Bioinformatics 35.1 (2019): 47-54. -->
</section>
<section id="psuedo-time-in-cell-lines-1" class="slide level2">
<h2>Psuedo-time in Cell lines</h2>
<div style="margin-top: 2.0em">
<figure style="text-align: center;">
<img src="images/PPCA_GPLVM_cell_data.png" alt="drawing" height="450">
<figcaption>
PPCA and GPLVM fitted to single-cell data.
</figcaption>
</figure>
</div>
<!-- [^5]:Ahmed, Sumon, Magnus Rattray, and Alexis Boukouvalas. "GrandPrix: scaling up the Bayesian GPLVM for single-cell data." Bioinformatics 35.1 (2019): 47-54. -->
</section>
<section id="motion-capture-data-6" class="slide level2">
<h2>Motion Capture Data <sup>1</sup></h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="fragment">
<div style="margin-top: 2.0em">
<ul>
<li>Each pose is represented by a 42 dimensional vector which consists of joint angles, and the position and orientation of the root of the kinematic chain.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>They then use a Gaussian process latent variable model to learn a low dimensional representation of the movement.</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>This has applications in animation from creating poses for interactive characters to doing real-time motion capture with missing markers.</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<div style="margin-top: 1.0em">
<figure style="text-align: center;">
<p><img src="images/motion_capture.png" alt="drawing" width="350"></p>
<figcaption>
Image from Grochow, Keith, et al.
</figcaption>
</figure>
</div>
</div>
</div>
<!-- ::: {.fragment}
::: {style="margin-top: 2.0em"}
- Each pose is represented by a 42 dimensional vector which consists of joint angles, and the position and orientation of the root of the kinematic chain.
:::
:::
::: {.fragment}
- They then use a Gaussian process latent variable model to learn a low dimensional representation of the movement.
:::
::: {.fragment}
- This has applications in animation from creating poses for interactive characters to doing real-time motion capture with missing markers.
::: -->
<aside><ol class="aside-footnotes"><li id="fn5"><p>Grochow, Keith, et al.&nbsp;“Style-based inverse kinematics.” ACM SIGGRAPH 2004 Papers. 2004. 522-531.</p></li></ol></aside></section>
<section id="motion-capture-data" class="slide level2">
<h2>Motion Capture Data</h2>
<div style="margin-top: 3.0em">
<figure style="text-align: center;">
<p><img src="images/grochow-et-al-2004.png" alt="drawing" width="800"></p>
<figcaption>
Grochow et al. <sup>1</sup> Human Motion Capture Latent Space.
</figcaption>
</figure>
</div>
<aside><ol class="aside-footnotes"><li id="fn6"><p>Grochow, Keith, et al.&nbsp;“Style-based inverse kinematics.” ACM SIGGRAPH 2004 Papers. 2004. 522-531.</p></li></ol></aside></section>
<section id="multi-output-gaussian-processes-7" class="slide level2">
<h2>Multi-output Gaussian Processes <sup>1</sup></h2>
<div class="columns">
<div class="column" style="width:45%;">
<div class="fragment">
<div style="margin-top: 3.0em">
<ul>
<li>Multiple output functions <span class="math inline">\(\{y_p\}_{p=1}^P\)</span>.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>In this model, the latent variable Gaussian process is assumed to have observed input dimensions <span class="math inline">\(\boldsymbol{x} \in \mathbb{R}^{D_x}\)</span> as well as unobserved one <span class="math inline">\(\boldsymbol{h} \in \mathbb{R}^{Q}\)</span>: <span class="math display">\[
y_p(\boldsymbol{x}) = f(\boldsymbol{x}, \boldsymbol{h}_p) + \epsilon.
\]</span></li>
</ul>
</div>
</div><div class="column" style="width:55%;">
<div class="fragment">
<div style="margin-top: 2.0em">
<figure style="text-align: center;">
<p><img src="images/LVMOGP_without_bayesian.png" alt="drawing" width="650"></p>
<figcaption>
Latent Variable Multi-output Gaussian Process
</figcaption>
</figure>
</div>
</div>
</div>
</div>
<!-- ::: {.fragment}
- In this model, the latent variable Gaussian process is assumed to have observed input dimensions as well as unobserved one $\boldsymbol{x} \in \mathbb{R}^{D_x}$:
$$
y_p(\boldsymbol{x}) = f(\boldsymbol{x}, \boldsymbol{h}_p) + \epsilon. 
$$

:::

::: {.fragment}
<figure style="text-align: center;">
  
<img src="images/LVMOGP_without_bayesian.png" alt="drawing" width="600"/> 

  <figcaption>Latent Variable Multi-output Gaussian Process </figcaption>
</figure>

::: -->
<aside><ol class="aside-footnotes"><li id="fn7"><p>Dai, Zhenwen, Mauricio Álvarez, and Neil Lawrence. “Efficient modeling of latent information in supervised learning using gaussian processes.” Advances in Neural Information Processing Systems 30 (2017).</p></li></ol></aside></section>
<section id="multi-output-gaussian-processes" class="slide level2">
<h2>Multi-output Gaussian Processes</h2>
<div class="columns">
<div class="column" style="width:45%;">
<div style="margin-top: 3.0em">
<div class="fragment">
<ul>
<li>This allows us to learn the correlation between different output functions, similar to the linear model of coregionalisation.</li>
</ul>
</div>
</div>
<div class="fragment">
<ul>
<li>This is useful for a range of applications including multi-fidelty and multi-task optimisation.</li>
</ul>
</div>
</div><div class="column" style="width:55%;">
<div style="margin-top: 2.0em">
<figure style="text-align: center;">
<p><img src="images/LVMOGP_without_bayesian.png" alt="drawing" width="650"></p>
<figcaption>
Latent Variable Multi-output Gaussian Process
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="multi-output-gaussian-processes-for-transfer-learning-8" class="slide level2">
<h2>Multi-output Gaussian Processes for Transfer Learning <sup>1</sup></h2>
<div class="columns">
<div class="column" style="width:50%;">
<div style="margin-top: 2.0em">
<figure style="text-align: center;">
<p><img src="images/hutter1.jpg" alt="drawing" width="650"></p>
<figcaption>
Simulated process data of the five historic products for the first experimental condition. The markers × represent measurements whichare taken after each full day.
</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<figure style="text-align: center;">
<p><img src="images/hutter2.jpg" alt="drawing" width="400"></p>
<figcaption>
One example of a product embedding. The dimensions of the embedding space do not carry any specific physical meaning rather the information isencoded in the distances between points.
</figcaption>
</figure>
</div>
</div>
<!-- 

<figure style="text-align: center;">
  
<img src="images/LVMOGP_without_bayesian.png" alt="drawing" width="600"/> 

  <figcaption>Latent Variable Multi-output Gaussian Process </figcaption>
</figure> -->
<!-- [^7]:Dai, Zhenwen, Mauricio Álvarez, and Neil Lawrence. "Efficient modeling of latent information in supervised learning using gaussian processes." Advances in Neural Information Processing Systems 30 (2017). -->
<aside><ol class="aside-footnotes"><li id="fn8"><p>Hutter, Clemens, et al.&nbsp;“Knowledge transfer across cell lines using hybrid Gaussian process models with entity embedding vectors.” Biotechnology and Bioengineering 118.11 (2021): 4389-4401.</p></li></ol></aside></section>
<section id="summary" class="slide level2">
<h2>Summary</h2>
<div class="fragment">
<div style="text-align: center; color: #107895; font-size: 1.2em">
<p>Gaussian processes can be used to learn low dimensional representations of high dimensional data.</p>
</div>
</div>
<div class="fragment">
<div style="text-align: center; color: #107895; font-size: 1.2em">
<p>Probabilistic Principal Component Analysis is equivalent to a Gaussian process latent variable model with a linear kernel.</p>
</div>
</div>
<div class="fragment">
<div style="text-align: center; color: #107895; font-size: 1.2em">
<p>This can be used for a wide range of applications such as grouping development stages in cell lines and determining latent dynamics of a system.</p>
</div>
</div>
</section>
<section id="useful-references" class="slide level2">
<h2>Useful References</h2>
<ul>
<li>Lawrence, Neil, and Aapo Hyvärinen. <a href="https://jmlr.org/papers/volume6/lawrence05a/lawrence05a.pdf">Probabilistic non-linear principal component analysis with Gaussian process latent variable models.</a> Journal of machine learning research 6.11 (2005).</li>
<li>Titsias, Michalis, and Neil D. Lawrence. <a href="https://proceedings.mlr.press/v9/titsias10a/titsias10a.pdf">Bayesian Gaussian process latent variable model.</a> Proceedings of the thirteenth international conference on artificial intelligence and statistics. JMLR Workshop and Conference Proceedings, 2010.</li>
<li>Ahmed, Sumon, Magnus Rattray, and Alexis Boukouvalas. <a href="https://academic.oup.com/bioinformatics/article/35/1/47/5047752">GrandPrix: scaling up the Bayesian GPLVM for single-cell data.</a> Bioinformatics 35.1 (2019): 47-54.</li>
<li>Grochow, Keith, et al.&nbsp;<a href="https://www.cs.utoronto.ca/~jacobson/seminar/grochow-et-al-2004.pdf">Style-based inverse kinematics.</a> ACM SIGGRAPH 2004 Papers. 2004. 522-531.</li>
<li>Dai, Zhenwen, Mauricio Álvarez, and Neil Lawrence. <a href="https://proceedings.neurips.cc/paper/2017/file/1680e9fa7b4dd5d62ece800239bb53bd-Paper.pdf">Efficient modeling of latent information in supervised learning using gaussian processes.</a> Advances in Neural Information Processing Systems 30 (2017). <!-- - Li, Ping, and Songcan Chen. [A review on Gaussian process latent variable models.](https://www.sciencedirect.com/science/article/pii/S2468232216300828) CAAI Transactions on Intelligence Technology 1.4 (2016): 366-376. --> <!-- - Lalchand, Vidhi, Aditya Ravuri, and Neil D. Lawrence. [Generalised GPLVM with stochastic variational inference.](https://proceedings.mlr.press/v151/lalchand22a.html) International Conference on Artificial Intelligence and Statistics. PMLR, 2022. --></li>
</ul>
</section>
<section id="the-end-and-some-useful-resources" class="slide level2">
<h2>The end! And Some Useful Resources…</h2>
<ul>
<li>Getting started with GPLVMs:
<ul>
<li><a href="https://pyro.ai/examples/gplvm.html">pyro jupyter notebook demonstration on the cell data</a></li>
<li><a href="https://docs.gpytorch.ai/en/latest/examples/045_GPLVM/Gaussian_Process_Latent_Variable_Models_with_Stochastic_Variational_Inference.html">pytorch jupyter notebook on GPLVMs</a></li>
<li><a href="https://gpflow.github.io/GPflow/2.4.0/notebooks/basics/GPLVM.html">tensorflow (GPflow) implementation of Bayesian GPLVMs</a></li>
</ul></li>
<li>Useful videos / slides:
<ul>
<li><a href="https://www.youtube.com/watch?v=fEWI6XDUGTo">Neil Lawrence: Bayesian GPLVMs and Deep GPs Video</a></li>
<li><a href="https://gpss.cc/gprs15b/assets/session3.pdf">Neil Lawrence: GPLVMs Slides</a></li>
</ul></li>
</ul>
<div style="text-align: center; font-size: 1em">
<p>My email: <code>r.sedgwick19@imperial.ic.ac.uk</code></p>
</div>
<!--
## Training the Bayesian GPLVM 

::: {.fragment}
- However the marginal like is now intractable as $H$ appears non-linearly in inverse of $K(H,H) + \sigma^2\mathbf{I}$.
:::

::: {.fragment}
- To solve this problem we are going to use variational inference.
:::

## Variational Inference

<!-- ::: {.fragment}

- What if we put a prior on $H$?
    $$\mathbf{h}_n \sim \mathcal{N}(\mathbf{\mu}_{h_n}, \Sigma_{h_n}).$$

::: -->
<!--
:::  {.fragment}
:::: {.columns}
::: {.column width="45%"}
- What if we put a prior on $H$? 
    $$
    p(\mathbf{y}_n |\mathbf{h}_n, \sigma^2) = \mathcal{N}(\mathbf{y}_n | \mathbf{0}, \mathbf{H}\mathbf{H}^T + \sigma^2\mathbf{I}).
    $$
:::
::: {.column width="45%"}
<img src="images/PCA_of.png" alt="drawing" width="500"/> 
:::

- The marginal likelihood becomes:
    $$p(Y|H, \mathbf{\theta}, \sigma_n) = \int p(Y|F, H, \mathbf{\theta}, \sigma_n) p(F|H, \mathbf{\theta}) p(H) dF dH.$$
- However this is intractable! as $H$ appears non-linearly in inverse of $K(H,H) + \sigma^2\mathbf{I}$
- To solve this problem we are going to use variational inference.

## Bayesian GPLVM VI 

- Our marginal likelihood is intractable:
$$p(Y|H, \mathbf{\theta}, \sigma_n) = \int p(Y|F, H, \mathbf{\theta}, \sigma_n) p(F|H, \mathbf{\theta}) p(H) dF dH.$$
- To address this we will use a variational distribution over the latent variables: 
$$q(H) = \prod^N_{n=1} \mathcal{N}(\mathbf{h_n}| M_n, S_n), $$
where $M_n$ and $S_n$ are the mean a variance of the variational distribution $q(\mathbf{h}_n)$.

## Bayesian GPLVM VI 

- A lower bound can then be expressed via Jensen's inequality ($\mathbb{E}[g(x)] \geq g(\mathbb{E}[x])$) as follows:
$$\begin{align*}
    \log p(Y|H) &= \log \int p(Y|H) p(H) dH \\
     & = \log \int p(Y|H) p(H) \frac{q(H)}{q(H)} dH \\
     & \geq \int q(H) \log \frac{p(Y|H) p(H)}{q(H)} dH.
\end{align*}$$
This can be rearranged to: 
$$
    \log p(Y|H) \geq \underbrace{\int q(H) \log p(Y|H) dH}_{\text{still intractable!}} - \underbrace{\int q(H) \log \frac{ q(H)}{p(H)} dH}_{\text{KL divergence}}.
    \label{eq:BGPLVM_LB}
$$

## Bayesian GPLVM VI 

- How do we make $\int q(H) \log p(Y|H) dH$ tractable?
- Add $N_U$ inducing variables $U = \in \mathbb{R}^{N_U \times D}$ evaluated at inducing point locations $Z = [\mathbf{z}_1, ..., \mathbf{z}_{N_U}]^T$ where $Z \in \mathbb{R}^{N_U \times Q}$
- This gives a joint Gaussian process prior between $F$ and $U$ as:
$$ p(F, U | H, Z) = p(F|U, H, Z) p(U |Z),
$$
- The joint distribution is then:
$$p(Y, F, U | H, Z) = p(Y| F)p(F|U, H, Z) p(U |Z). $$
- Conditioning on $U$ decouples $F$ from each other meaning: 
 $$p(Y|H) = \prod_{n=1}^N p(\mathbf{y_n}|H).$$

## Bayesian GPLVM VI 

Include illistration of inducing variables / points

- How do we make $\int q(H) \log p(Y|H) dH$ tractable?
- Add $N_U$ inducing variables $U = \in \mathbb{R}^{N_U \times D}$ evaluated at inducing point locations $Z = [\mathbf{z}_1, ..., \mathbf{z}_{N_U}]^T$ where $Z \in \mathbb{R}^{N_U \times Q}$
- This gives a joint Gaussian process prior between $F$ and $U$ as:
$$ p(F, U | H, Z) = p(F|U, H, Z) p(U |Z),$$
- The joint distribution is then:
$$ p(Y, F, U | H, Z) = p(Y| F)p(F|U, H, Z) p(U |Z). $$

## Bayesian GPLVM VI 

- Conditioning on $U$ decouples $F$ from each other meaning: 
 $$p(Y|H) = \prod_{n=1}^N p(\mathbf{y_n}|H).$$
- This means $p(Y)$ can be found by marginalising out $F$ and $U$ for any value of $Z$.

## Bayesian GPLVM VI 

- By introducing a second variational distribution:
 $$q(F, U) = p(F| U, H) q(U),$$
 we can get a tractable lower bound:
 $$
 \begin{align*}
    \log p(Y|H) &= \log  \int p(Y|F) p(F |U, H) p(U) dH \\
    &= \log \int p(Y|F) p(F |U, H) p(U) \frac{p(F |U, H) q(U)}{p(F |U, H) q(U)} dH \\
    &= \log \mathbb{E} \left[\frac{p(Y| F) p(F |U, H) p(U)}{p(F |U, H) q(U)}\right]_{p(F |U, H) q(U)} \\
    &\geq \int p(F |U, H) q(U) \log \left( \frac{p(Y|F) p(U)}{q(U)}  \right) ~ dF dU \\
    &= \int q(U) \left( \int p(F| U, H) \log p(Y|F) ~d(F) + \log\frac{p(U)}{q(U)} \right) ~ dU, 
  \end{align*}
  $$

## Bayesian GPLVM VI 

- Putting this back into our lower bound we get: 
$$\begin{multline*}
    \log p(Y|H) \geq  \int q(H) \Biggl[ \int q(U) \Biggl( \int p(F| U, H) \log p(Y|F) ~d(F) \\  + \log\frac{p(U}{q(U)} \Biggr) ~ d(U)    - \log \frac{ q(H)}{p(H)} \Biggr] dH.
\end{multline*}
$$

- This is tractable! We therefore can maximise this lower bound instead of maximising the marginal likelihood. 
 -->
<!-- 
## Assumptions for Causal Discovery



::: {.notes}
- Main question of interest: How far can we get with observational data alone

- Let's make an assumption on the model that generated the data and fit that model

- Only one causal direction really FITS the data

- Basic strategy: data + assumptions
:::

::: {.fragment}
::: {style="text-align: center; color: #107895; font-size: 1.2em; font-weight: 900em"}
How far can we get with observational data alone?
:::
:::

::: {.fragment .fade-in-then-out}
::: {style="text-align: center"}

Assumptions: $E = f(C) + N$ (ANM)

::: {layout-ncol=2}
![$X \to Y$](images/gp_plot_causal:True.png)

![$Y \to X$](images/gp_plot_causal:False.png)



:::
:::
:::

::: {.fragment}
::: {style="text-align: center; margin-top: -10em"}
Observational data + assumptions $\to$ Causal structure
:::
:::



## Need for flexible assumptions



::: {.notes}
- A problem with assumptions are that they can be restrictive

- What assumption can we make for data like this?

- Let's try and previous ANM

- Can we really trust this answer?
:::

Assumptions that allow for identifiability can be restrictive

::: {.fragment .fade-in-then-out}
:::{style="text-align: center"}

:::{.absolute  top=150 left=400}
Assumption: ?
:::

![](images/gplvm_data_plot.png){.absolute left=150 top=250 height="60%" width=65%}
:::
:::

::: {.fragment}
:::{style="text-align: center"}

Assumption: $E = f(C) + N$ (ANM)

::: {layout-ncol=2}
![$X \to Y$](images/gp_fit_on_gplvm_causal.png)

![$Y \to X$](images/gp_fit_on_gplvm_anticausal.png)
:::
:::
:::

## Need for flexible assumptions

::: {.notes}
- This is the main question we will address today
:::


::: {.absolute top=200}
::: {style="text-align: center; color: #107895; font-size: 1.4em; font-weight: 900em"}
Can we do causal discovery with more flexible assumptions?
:::
:::

## Back to basics: Factorisation and interventions

Causal model $X \to Y$ specifies a joint distribution $\Pi(X, Y)$


::: {.fragment}
Causal model defines a data generation process:
$$
\begin{align}
X :=& \  f_x(N_x) \\
Y :=& \ f_y(X, N_y)
\end{align}
$$
:::

::: {.fragment}
- $\Pi(X, Y) = \Pi(X) \Pi(Y|X)$: Causal factorisation
- $\Pi(X, Y) = \Pi(Y) \Pi(X|Y)$: Anticausal factorisation
:::

::: {.fragment}
- Intervention on $X$: Changing $f_x$ or $N_x$
- Intervention on $Y$: Changing $f_y$ or $N_y$
:::






## Back to basics: ICM

$X \to Y$:
$$
\begin{align}
X :=& \  f_x(N_x) \\
Y :=& \ f_y(X, N_y)
\end{align}
$$


::: {.fragment}


::: {style="margin-top: -1em"}
- Changing $X$ to $\Pi'(X)$ leaves $\Pi(Y|X)$ unchanged
- Changing $X$ to $\Pi'(X)$ [does not]{style="color:#42affa"} leave $\Pi(Y) = \sum_X \Pi(X) \Pi(Y|X)$ unchanged
:::

:::

::: {.fragment}
- The [causal factorisation]{style="color:#42affa"} is special.
  - Causal mechanisms are ["independent"]{style="color:#42affa"} while anticausal may not be
-  Also known as the [Independent causal mechanism]{style="color:#42affa"}  (ICM) assumption.
:::



::: {.fragment}
E.g. Altitude $\to$ Temperature: changing ensemble of altitudes $\Pi(\text{Altitude})$ will not change $\Pi(\text{Temperature | Altitude})$.

:::

## Parametrising Causal models

We want to make sure our parametrisation follows ICM

$$
\begin{gather}
    p(\bfx, \bfy|\phi, \mathbf{\theta}, \xcausey) = p(\bfx|\phi, \xcausey)p(\bfy|\bfx,\mathbf{\theta}, \xcausey)\,, \\
    p(\bfx, \bfy|\phi, \mathbf{\theta}, \ycausex) = p(\bfy|\phi, \ycausex)p(\bfx|\bfy,\mathbf{\theta}, \ycausex)\,,
\end{gather}
$$

::: {.fragment}

::: {.absolute height="90" width="510" top=170 left=440}
::: {.box}
:::
:::

::: {.absolute top=110 left=560}
::: {style="color: #D24000; font-size: 0.8em; font-style:italic"}
Causal factorisation
:::
:::

:::

::: {.fragment}
- Marginals are chosen from $\setmarginals$ and conditionals are chosen from $\setconditionals$

::: {style="margin-top: -1em"}
$$
\begin{align}
\setmarginals = \{p(\cdot| \phi) \suchthat \phi \in \Phi \} && \setconditionals  = \{p(\cdot| \cdot, \mathbf{\theta}) \suchthat \mathbf{\theta} \in \Theta  \}
\end{align}
$$
:::

::: {.fragment}
- Allows for parts of the [**correct**]{style="color: #003E74"} model to be reused under interventions
:::

:::





## Why doesn't fitting models work?

:::{.fragment}
- Given data, we can find $\phi, \mathbf{\theta}$ by maximising the likelihood
:::

:::{.fragment}
- For [flexible models]{style="color:#42affa"}, we need $\setmarginals$ and $\setconditionals$ to be large enough so that we can predict well on any given dataset
:::

:::{.fragment}
- Can we distinguish $\xcausey$ and $\ycausex$?
:::


:::{.fragment}
:::{style="color:#42affa; text-align: center"}
**No**
:::
:::

::: {.fragment}
::: {style="font-size: 0.8em; text-align: center"}
$$
    \max_{\mathbf{\theta}, \phi}  p(\bfx|\phi, \xcausey)p(\bfy|\bfx,\mathbf{\theta}, \xcausey) = \max_{\mathbf{\theta}, \phi}  p(\bfy|\phi, \ycausex)p(\bfx|\bfy,\mathbf{\theta}, \ycausex)
$$
:::

::: {style="font-size: 0.8em; text-align: center"}
$$
    P(X)P(Y|X) = P(Y)P(X|Y)
$$
:::


:::



## Why doesn't fitting models work?

<br>

::: {style="font-size: 0.8em; text-align: center"}
$$
   \underbrace{\max_{\mathbf{\theta}, \phi}  p(\bfx|\phi, \xcausey)p(\bfy|\bfx,\mathbf{\theta}, \xcausey)}_{\in \setmarginals \times \setconditionals} =  \underbrace{\max_{\mathbf{\theta}, \phi} p(\bfy|\phi, \ycausex)p(\bfx|\bfy,\mathbf{\theta}, \ycausex)}_{\in \setmarginals \times \setconditionals }
$$
:::

::: {.absolute left=200 height="500" width="700"}

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-2-1.png){width=960}
:::
:::

:::



## Why do restrictions work?


::: {.fragment}
:::{style="text-align: center"}

$\setconditionals =$ Additive non-linear model (ANM)

<br>

::: {.absolute left=100 top=128}
$p(\bfy|\bfx,\mathbf{\theta}, \xcausey) \in \setconditionals$
:::


::: {.absolute left=620 top=128}
$p(\bfx|\bfy,\mathbf{\theta}, \ycausex) \notin \setconditionals$
:::



::: {.absolute top=228}
::: {layout-ncol=2}
![](images/gp_plot_causal:True.png)

![](images/gp_plot_causal:False.png)
:::
:::

::: {.absolute top=550 left=100}
::: {style="font-size: 0.8em; text-align: center"}
$$
    \underbrace{\max_{\mathbf{\theta}, \phi}  p(\bfx|\phi, \xcausey)p(\bfy|\bfx,\mathbf{\theta}, \xcausey)}_{\in \setmarginals \times \setconditionals} > \underbrace{\max_{\mathbf{\theta}, \phi} p(\bfy|\phi, \ycausex)p(\bfx|\bfy,\mathbf{\theta}, \ycausex)}_{\notin \setmarginals \times \setconditionals}
$$
:::
:::

:::
:::


## Why do restrictions work?

::: {.fragment}
::: {.absolute left=200 height="500" width="700"}
::: {style="text-align: center"}

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-3-1.png){width=960}
:::
:::


::: {.absolute top=280 left=-100}
::: {style="font-size: 0.8em; text-align: center"}
$$
    \underbrace{\max_{\mathbf{\theta}, \phi}  p(\bfx|\phi, \xcausey)p(\bfy|\bfx,\mathbf{\theta}, \xcausey)}_{\in \setmarginals \times \setconditionals} > \underbrace{\max_{\mathbf{\theta}, \phi} p(\bfy|\phi, \ycausex)p(\bfx|\bfy,\mathbf{\theta}, \ycausex)}_{\notin \setmarginals \times \setconditionals}
$$
:::
:::


:::
:::
:::

::: {.absolute top=480}
::: {.fragment}
- Restricting the class of models allows for maximum likelihood to identify causal direction
:::
:::


::: {.absolute top=600}
::: {.fragment}
- This also restricts the datasets that you can model!

::: {.absolute top=-450 left=530}
::: {style="color: #D24000; font-size: 60px"}
**?**
:::
:::

:::
:::

## Why do restrictions work?

::: {.fragment}
:::{style="text-align: center"}



$\setconditionals =$ Additive non-linear model (ANM)

<br>

::: {.absolute left=100 top=158}
$p(\bfy|\bfx,\mathbf{\theta}, \xcausey) \notin \setconditionals$
:::


::: {.absolute left=620 top=158}
$p(\bfx|\bfy,\mathbf{\theta}, \ycausex) \notin \setconditionals$
:::

::: {.absolute top=258}
::: {layout-ncol=2}
![$X \to Y$](images/gp_fit_on_gplvm_causal.png)

![$Y \to X$](images/gp_fit_on_gplvm_anticausal.png)
:::
:::

:::
:::

## Causal Discovery: Complexity


:::{.fragment}
Another line of work has relied on the observation that the causal factorisation is "simple"

- E.g. Only causal factorisation can fit in "simple" model
:::

:::{.fragment}
Different ways of measuring simplicity

- Kolmogorov complexity
- Minimum description length
:::
:::{.fragment}

Bayesian formulation has automatic [complexity control]{style="color:#42affa"}

- Relies on similar assumptions
- Has the same lack of guarantees
:::

## Bayesian Model selection

[**Bayes**:]{style="color:#003E74"} Find the posterior of each causal model

::: {.fragment}
$$
\begin{align}
    p(\xcausey|\data) = \frac{p(\data|\xcausey)p(\xcausey)}{p(\data|\xcausey)p(\xcausey) + p(\data|\ycausex)p(\ycausex)} \,
\end{align}
$$
:::


::: {.fragment}
We can summarise

$$
\begin{align}
    \log \frac{p(\xcausey|\data)}{p(\ycausex|\data)} =
    \log \frac{p(\data|\xcausey)p(\xcausey)}{p(\data|\ycausex)p(\ycausex)} \,
\end{align}
$$
:::

::: {.fragment}
Unsure of causal direction: $p(\xcausey) = p(\ycausex) = 0.5$
:::

::: {.fragment}
Important terms are the [**Marginal likelihoods**]{style="color:#D24000"}:
$\log \frac{p(\data| \xcausey)}{p(\data| \ycausex)}$


:::


## Marginal Likelihoods

Marginal likelihoods can be calculated:

::: {style="font-size: 30px"}
$$
\begin{align}
    p(\bfx, \bfy|\xcausey) = \iint p(\bfx | \phi, \xcausey) p(\bfy|\bfx, \mathbf{\theta}, \xcausey) p(\phi,\mathbf{\theta}| \xcausey) \calcd\phi \calcd\mathbf{\theta} \\
    p(\bfx, \bfy|\ycausex) = \iint p(\bfy | \phi, \ycausex) p(\bfx|\bfy, \mathbf{\theta}, \ycausex) p(\phi, \mathbf{\theta}| \ycausex)\calcd\phi \calcd \mathbf{\theta}
\end{align}
$$
:::

::: {.fragment}
- We must specify priors $p(\phi, \mathbf{\theta}| \cdot)$
:::
::: {.fragment}
- ICM $\implies$ $\phi \perp \!\!\! \perp \mathbf{\theta}$
:::


::: {.fragment}
- Priors represent our [assumptions]{style="color:#42affa"}
  - Can place zero mass in areas to mimic restricting $\setmarginals, \setconditionals$
:::

::: {.fragment}
- Real example of choice of prior later
:::

## Can BMS distinguish causal models?

For both causal models to have the same score for all datasets


$$
p(\bfx, \bfy|\xcausey) = p(\bfx, \bfy|\ycausex) \ \forall \bfx, \bfy
$$

... a model must factorise according to [**both**]{style="color:#003E74"} graphical models

::: {style="text-align: center"}
::: {layout-ncol=2}


::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-4-1.png){width=960}
:::
:::

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-5-1.png){width=960}
:::
:::





:::
:::

::: {.absolute top=500}
::: {.fragment .fade-in-then-out}
::: {style="font-size: 26px"}
$$
p(\bfx | \phi, \xcausey) p(\bfy|\bfx, \mathbf{\theta}, \xcausey) p(\phi) p(\mathbf{\theta})
$$
:::
:::
:::

::: {.absolute top=500}
::: {.fragment .fade-in-then-out}
::: {style="font-size: 26px"}
$$
p(\bfx | \phi, \xcausey) p(\bfy|\bfx, \mathbf{\theta}, \xcausey) p(\phi) p(\mathbf{\theta}) = p(\bfy | \phi, \mathbf{\theta}, \xcausey) p(\bfx|\bfy, \phi, \mathbf{\theta}, \xcausey) p(\phi) p(\mathbf{\theta})
$$
:::
:::
:::

::: {.absolute top=500}
::: {.fragment .fade-in}
::: {style="font-size: 26px"}
$$
p(\bfx | \phi, \xcausey) p(\bfy|\bfx, \mathbf{\theta}, \xcausey) p(\phi) p(\mathbf{\theta}) = p(\bfy | \phi,  \xcausey) p(\bfx|\bfy, \mathbf{\theta}, \xcausey) p(\phi) p(\mathbf{\theta})
$$
:::
:::
:::

::: {.absolute top=570}
::: {.fragment}
- E.g. If $\setmarginals, \setconditionals$ are Normalised linear Gaussian
:::
:::

## Can BMS distinguish causal models?



::: {style="text-align: center"}
 If $\setconditionals, \setmarginals$ are  [**restricted**]{style="color: #003E74"} such that maximum likelihood can distinguish, can BMS?
:::

::: {.fragment}
:::{style="text-align: center; color: #42affa"}
**Yes**
:::
:::


::: {.fragment}
::: {style="text-align: center"}
::: {layout-ncol=2}


::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-6-1.png){width=960}
:::
:::

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-7-1.png){width=960}
:::
:::



:::
:::
:::


::: {.fragment}

- Retains identifiability guarantees for models like ANM

:::


::: aside
Full theorems and assumptions needed in the paper
:::

## Can BMS distinguish causal models?


::: aside
Full theorems and assumptions needed in the paper
:::

::: {style="text-align: center"}
 If $\setconditionals, \setmarginals$ are [**flexible**]{style="color: #003E74"} such that maximum likelihood [**cannot**]{style="color: #D24000"} distinguish, can BMS?
:::

::: {.fragment}
:::{style="text-align: center; color: #42affa"}
**Yes!**
:::
:::

::: {.fragment}
::: {style="text-align: center"}
::: {layout-ncol=2}


::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-8-1.png){width=960}
:::
:::

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-9-1.png){width=960}
:::
:::



:::
:::
:::

::: {.fragment}

::: {.absolute top=170 left=700}
::: {style="color: #D24000; font-size: 20px"}
$$
\sum_{\data} p(\data| \cdot) = 1
$$
:::
:::

::: {.absolute top=170 left=150}
::: {style="color: #D24000; font-size: 20px"}
$$
\sum_{\data} \max_{\mathbf{\theta}, \phi} p(\data| \mathbf{\theta}, \phi, \cdot) < \infty
$$
:::
:::

:::

::: {.fragment}

- BMS has an opinion for [**flexible**]{style="color: #003E74"} $\setmarginals, \setconditionals$ when maximum likelihood is indifferent

:::


## Does BMS find the correct causal direction?

Previous result said nothing about finding the  [**correct**]{style="color: #003E74"} causal model.

::: {.fragment}

- Assume datasets are sampled from $\Pi(X,Y)$
- Assume one of our models is correct (assumptions match data generation)
  - E.g. $p(X, Y| \xcausey) = \Pi(X,Y)$


:::


::: {.fragment}
Decision rule:

::: {style="font-size: 30px"}
$$
\begin{align}
\mathcal{M}^* = \begin{cases} \xcausey \text{ if } p(\data| \xcausey) > p(\data| \ycausex) \\ \ycausex \text{ if } p(\data| \xcausey) < p(\data| \ycausex) \end{cases}
\end{align}
$$
:::
:::



## Does BMS find the correct causal direction?




We focus on the probability of error


::: {style="font-size: 24px"}
$$
\begin{align}
    P(E| \xcausey) = \int_{\mathcal{R}_Y} p(\data| \xcausey ) \calcd \data \,, && \mathcal{R}_Y = \{\data \suchthat p(\data|\ycausex) > p(\data|\xcausey)\} \,.
\end{align}
$$
:::
- Can be calculated numerically for our choice of model




Probability of error:

::: {style="font-size: 30px"}
$$
\begin{align}
    P(E) = \frac{1}{2} (1-  \underbrace{\text{TV}[P_{\data}(\cdot| \xcausey ),  P_{\data}(\cdot| \ycausex )])}_{\text{Total variation distance between dataset densities}},
\end{align}
$$
:::

::: {.fragment .fade-in-then-out}

::: {.absolute top=600}
- Restrictions imply $\text{TV} = 1 \implies P(E) = 0$
  - Retains notion of identifiability
:::


::: {.absolute top=550 height="200" width="300" left=800}

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-10-1.png){width=960}
:::
:::

:::
:::

::: {.fragment .fade-in}

::: {.absolute top=600}
- Flexibility $\implies$ may be non-zero probability of error
:::

::: {.absolute top=550 height="200" width="300" left=800}

::: {.cell}
::: {.cell-output-display}
![](presentation_files/figure-revealjs/unnamed-chunk-11-1.png){width=960}
:::
:::

:::

:::


## Does BMS find the correct causal direction? (Misspecification)

What if we don't know the true data generating process (true data generating prior) ?

$$
\begin{align}
|\underbrace{\Pi(\text{Error})}_{\text{True probability of error}} - \underbrace{P(\text{Error})}_{\text{Model probability of error}}| \leq
\text{TV}[  \Pi_{\mathcal{D}}(\cdot| X \to Y), P_{\mathcal{D}}(\cdot| \mathcal{M}_{X \to Y})]
\end{align}
$$

::: {.fragment}
- Correct priors are hard to verify
:::


::: {.fragment}
- Above shows that it is not necessary to exactly get the priors right
:::

## Method

We want a flexible model that has good predictive ability regardless of the dataset given

::: {.fragment}
- $\setmarginals$: GPLVM, $\setconditionals$: Conditional GPLVM
:::

::: {.fragment}
$$
\begin{align}
p(y_i| x_i, f_y, \xcausey) &= \int p(y_i| x_i, f_y, w_i \xcausey) p(w_i) dw_i \\
p(x_i| f_x, \xcausey) &= \int p(x_i| f_x, w_i \xcausey) p(w_i) dw_i
\end{align}
$$

- Infinite mixture of Guassians can model:
  - Non -Gaussian likelihoods
  - Heteroscedastic noise
:::

## Flexibility of GPLVM

::: {.absolute height="300" width="800" left=100}
::: {style="text-align: center"}
::: {layout-ncol=2}
![$X \to Y$](images/gplvm_plot_nofullcov_causal:True.png)

![$Y \to X$](images/gplvm_plot_nofullcov_causal:False.png)

:::
:::
:::


::: {.fragment}
::: {.absolute height="300" width="800" top=375 left=100}
::: {style="text-align: center"}
::: {layout-ncol=2}
![$X \to Y$](images/gplv_data_gplvm_fit_no_full_cov_causal:True.png)

![$Y \to X$](images/gplv_data_gplvm_fit_no_full_cov_causal:False.png)

:::
:::
:::
:::

## Results

- How does the flexible model work with datasets that are generated from identifiable models?
- How does it compare against the identifiable models themselves?

::: {.fragment}

::: {style="text-align: center"}

Numbers are ROC AUC (higher is better)

| Methods | ANM |
 | -------| ----|
| Gaussian Process | 100.0 |
 |  GPLVM | 100.0 |

:::

- Flexibility does not necessarily hurt performance for idenitifiable models
:::

## Results

- Empirical evaluation on varying data generating processes

| Methods |   Cha | Multi | Net | Gauss | Tueb |
| :----:| :----:| :----:| :----:| :---:| :----:|
 | LiNGAM  | 57.8 | 62.3 | 3.3 | 72.2 | 31.1 |
 | ANM   | 43.7 |25.5 | 87.8 | 90.7 | 63.9 |
 | PNL  | [78.6]{style="text-decoration: underline;"} | 51.7 | 75.6 | 84.7 | 73.8 |
  | IGCI  | 55.6 | 77.8 | 57.4 | 16.0 | 63.1  |
  |SLOPPY  | 60.1 | 95.7 | 79.3 | 71.4 | 65.3 |
  | RECI  | 59.0 | 94.7 | 66.0 | 71.0  | 70.5 |
 | CGNN   | 76.2 | 94.7 | 86.3 | 89.3 | [76.6]{style="text-decoration: underline;"} |
 | GPI   | 71.5 | 73.8 | 88.1 | 90.2 | 70.6  |
  |CDCI | 72.2 | [96.0]{style="text-decoration: underline;"} |  [94.3]{style="text-decoration: underline;"} | [**91.8**]{style="text-decoration: underline;"} | - |
  | **GPLVM** | **81.9** | **97.7** | **98.9** | 89.3 | **78.3**|


::: {.fragment .fade-in-then-out}

::: {.absolute height="290" width="130" top=200 left=90}
::: {.box}
:::
:::

::: {.absolute top=250 left=-80}
[Strong]{style="color: #D24000"}
<br>
[assumptions]{style="color: #D24000"}
:::

:::


::: {.fragment .fade-in-then-out}

::: {.absolute height="150" width="130" top=500 left=90}
::: {.box}
:::
:::

::: {.absolute top=550 left=-80}
[Complexity]{style="color: #D24000"}
<br>
[control]{style="color: #D24000"}
:::

:::


::: {.fragment .fade-in-then-out}

::: {.absolute height="50" width="900" top=650 left=90}
::: {.box}
:::
:::



:::

## Why not better?

We numerically calculate that for our model $P(E) \approx 0$. Why is the error higher?

- Model not flexible enough
- Prior could be misspecified
- Approximations in inference introducing errors


## Conclusion

- Causal discovery requires assumptions
- Bayesian model selection provides a way of encoding assumptions
- Allows for use of flexible models and specification of realistic assumptions
- This leads to good performance
- Specific choice of assumptions still depends on the problem at hand

::: {style="text-align: center"}
Causal representation $\to$ Learn variables + Causal structure $\to$ Need flexible models!
::: -->
<div class="quarto-auto-generated-content">
<div class="footer footer-default">

</div>
</div>
</section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="presentation_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="presentation_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1150,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>
